### Input/output options

Define where the pipeline should find input data and save output data.

#### `--input`

- **Required**
- Type: string

Sample sheet with sample names and paths to reads.


#### `--platform`

- Optional
- Type: string

NGS platform used to sequence the samples.


#### `--ref_db`

- Optional
- Type: string

Provide fasta database of interesting reference sequences


#### `--outdir`

- Optional
- Type: string
- Default: `./results`

The output directory where the results will be saved.


### Nextclade options

Nextclade dataset options.

#### `--with_nextclade`

- Optional
- Type: boolean
- Default: `true`

Run and update Nextclade (only runs on Nanopore data)


#### `--nextclade_dataset`

- Optional
- Type: string
- Default: `flu_h1n1pdm_ha`

Nextclade dataset to use


#### `--nextclade_reference`

- Optional
- Type: string
- Default: `CY121680`

Dataset reference to use


#### `--nextclade_tag`

- Optional
- Type: string
- Default: `2023-04-02T12:00:00Z`

dataset tag


### Variant calling options

Various options for the variant calling branch of the pipeline.

#### `--variant_caller`

- Optional
- Type: string
- Default: `clair3`

Variant caller (Clair3 or Medaka)


#### `--medaka_variant_model`

- Optional
- Type: string
- Default: `r941_prom_hac_variant_g507`

[Medaka](https://github.com/nanoporetech/medaka) Medaka model for final variant calling from phased reads


#### `--medaka_snp_model`

- Optional
- Type: string
- Default: `r941_prom_hac_snp_g507`

[Medaka](https://github.com/nanoporetech/medaka) Medaka model for initial SNP calling from mixed reads prior to phasing


#### `--clair3_variant_model`

- Optional
- Type: string
- Default: `r941_prom_sup_g5014`

[Clair3](https://github.com/HKU-BAL/Clair3) Clair3 variant model


#### `--minor_allele_fraction`

- Optional
- Type: number
  - Maximum: 1.0
- Default: `0.25`

Minor variant allele frequency/fraction.


#### `--major_allele_fraction`

- Optional
- Type: number
  - Maximum: 1.0
- Default: `0.75`

Major variant allele frequency/fraction. Only major variant alleles are used for generating a consensus sequence.


#### `--low_coverage`

- Optional
- Type: integer
- Default: `10`

Low coverage depth threshold. Consensus sequence positions with less than this coverage depth will be masked with `N`.


### Nanopore options

Options exclusive to running the pipeline on Nanopore data using the ARTIC fieldbioinformatics pipeline.

#### `--min_sample_reads`

- Optional
- Type: integer
- Default: `100`

Minimum number of raw reads required per sample in order to be considered for the downstream processing steps.


### IRMA assembly options

#### `--irma_module`

- Optional
- Type: string

IRMA module to use for analysis.


#### `--keep_ref_deletions`

- Optional
- Type: boolean
- Default: `true`

Set "DEL_TYPE=NNN" to keep deletions to reference sequence as N characters in consensus.


#### `--skip_irma_subtyping_report`

- Optional
- Type: boolean
- Default: `true`

Skip the output of subtyping report generated by IRMA consensus sequences.


### Nanopore/Illumina options

Options common to both the Nanopore and Illumina workflows in the pipeline.

#### `--skip_mosdepth`

- Optional
- Type: boolean

Skip genome-wide and amplicon coverage plot generation from mosdepth output.


### Mistmatch report options

Reporting option.

#### `--min_aln_length`

- Optional
- Type: integer
- Default: `700`

Minimum alignment length of nucleotide BLAST results to consider for getting mismatch report from BLAST result against reference database. We normally set this number around the smallest genome segment (Segment 8).


### H/N subtyping options

Hemaglutinin and neuraminase subtype prediction options

#### `--pident_threshold`

- Optional
- Type: number
  - Maximum: 1
- Default: `0.85`

Minimum % identity of nucleotide BLAST results to consider for determining H/N subtypes.


#### `--max_top_blastn`

- Optional
- Type: integer
  - Minimum: 1
- Default: `3`

Maximum of top blastn result reported


#### `--ncbi_influenza_fasta`

- Optional
- Type: string
- Default: `https://ftp.ncbi.nih.gov/genomes/INFLUENZA/influenza.fna.gz`

Path/URL to NCBI Influenza DB sequences FASTA file.


#### `--ncbi_influenza_metadata`

- Optional
- Type: string
- Default: `https://ftp.ncbi.nih.gov/genomes/INFLUENZA/genomeset.dat.gz`

Path/URL to NCBI Influenza DB metadata file.


### Generic options

Less common options for the pipeline, typically set in a config file.

#### `--help`

- Optional
- Type: boolean

Display help text.


#### `--publish_dir_mode`

- Optional
- Type: string
- Default: `copy`

Method used to save pipeline results to output directory.

> **NOTE:** The Nextflow `publishDir` option specifies which intermediate files should be saved to the output directory. This option tells the pipeline what method should be used to move these files. See [Nextflow docs](https://www.nextflow.io/docs/latest/process.html#publishdir) for details.


#### `--monochrome_logs`

- Optional
- Type: boolean

Do not use coloured log outputs.

> **NOTE:** Set to disable colourful command line output and live life in monochrome.


#### `--tracedir`

- Optional
- Type: string
- Default: `${params.outdir}/pipeline_info`

Directory to keep pipeline Nextflow logs and reports.


#### `--enable_conda`

- Optional
- Type: boolean

Run this workflow with Conda. You can also use '-profile conda' instead of providing this parameter.


#### `--singularity_pull_docker_container`

- Optional
- Type: boolean

Instead of directly downloading Singularity images for use with Singularity, force the workflow to pull and convert Docker containers instead.


#### `--validate_params`

- Optional
- Type: boolean
- Default: `True`

Boolean whether to validate parameters against the schema at runtime


#### `--show_hidden_params`

- Optional
- Type: boolean

Show all params when using `--help`

> **NOTE:** By default, parameters set as _hidden_ in the schema are not shown on the command line when a user runs with `--help`. Specifying this option will tell the pipeline to show all parameters.


### Slurm scheduler options

#### `--slurm_queue`

- Optional
- Type: string

Slurm queue/partition to submit pipeline jobs to


#### `--slurm_queue_size`

- Optional
- Type: integer
- Default: `100`

Slurm queue size. Max number of jobs to queue at once.


### Max job request options

Set the top limit for requested resources for any single job.

#### `--max_cpus`

- Optional
- Type: integer
- Default: `16`

Maximum number of CPUs that can be requested    for any single job.

> **NOTE:** Use to set an upper-limit for the CPU requirement for each process. Should be an integer e.g. `--max_cpus 1`


#### `--max_memory`

- Optional
- Type: string
- Default: `128.GB`

Maximum amount of memory that can be requested for any single job.

> **NOTE:** Use to set an upper-limit for the memory requirement for each process. Should be a string in the format integer-unit e.g. `--max_memory '8.GB'`


#### `--max_time`

- Optional
- Type: string
- Default: `240.h`

Maximum amount of time that can be requested for any single job.

> **NOTE:** Use to set an upper-limit for the time requirement for each process. Should be a string in the format integer-unit e.g. `--max_time '2.h'`


